pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/v2/opt/rules/LeafStageWorkerAssignmentRule.java

```diff
@@ -292,6 +303,9 @@ static TableScanWorkerAssignmentResult attemptPartitionedDistribution(String tab
           }
         }
       }
+      if (invalidSegmentsByInferredPartition.containsKey(partitionNum)) {
```

Comment:
Would it be possible that the invalidSegmentsByInferredPartition's keyset is the superset of numPartition?

Reply:
Great catch. We should take modulus in that case. (though this case is very unlikely, since it would mean that we have a scenario like 256 kafka partitions but 64 data partitions).